src/mlpack/core/metrics/bleu_impl.hpp

```diff
@@ -0,0 +1,200 @@
+ElemType BLEU<ElemType, PrecisionType>::Evaluate(
+    const ReferenceCorpusType& referenceCorpus,
+    const TranslationCorpusType& translationCorpus,
+    const bool smooth)
+{
+  // WordVector is a string container type.
+  // Also, TranslationCorpusType is an array of such containers.
+  typedef typename TranslationCorpusType::value_type WordVector;
+
+  // matchesByOrder: It catches how many times sequence of a particular order
+  // is encountered in both reference corpus and translation corpus.
+  std::vector<size_t> matchesByOrder(maxOrder, 0);
+
+  // possibleMatchesByOrder: It tracks how many possible matches can be in the
+  // translation corpus.
+  std::vector<size_t> possibleMatchesByOrder(maxOrder, 0);
+
+  // referenceLength: It is the sum of minimum length of the paragraph from
+  // various documents.
+  // translationLength: It is the sum of length of each paragraphs.
+  referenceLength = 0, translationLength = 0;
+
+  auto refIt = referenceCorpus.cbegin();
+  auto trIt = translationCorpus.cbegin();
+  for (; refIt != referenceCorpus.cend(), trIt != translationCorpus.cend();
+      ++refIt, ++trIt)
+  {
+    size_t min = std::numeric_limits<size_t>::max();
+    for (const auto& t : *refIt)
+    {
+      if (min > t.size())
+      {
+        min = t.size();
+      }
+    }
+
+    if (min == std::numeric_limits<size_t>::max())
+      min = 0;
+
+    referenceLength += min;
+    translationLength += trIt->size();
+
+    // mergedRefNGramCounts: It accumulates all the similar n-grams from
+    // various references or documents, so that there is no repetition of
+    // any key (sequence of order n).
+    std::map<WordVector, size_t> mergedRefNGramCounts;
+    for (const auto& t : *refIt)
+    {
+      // ngram: It holds the n-grams of each document/reference.
+      const std::map<WordVector, size_t> ngrams = GetNGrams(t);
+      for (auto it = ngrams.cbegin(); it != ngrams.cend(); ++it)
+      {
+        mergedRefNGramCounts[it->first] = std::max(it->second,
+            mergedRefNGramCounts[it->first]);
+      }
+    }
+    // translationNGramCounts: It extracts the n-grams of the generated text
+    // sequence.
+    const std::map<WordVector, size_t> translationNGramCounts
+        = GetNGrams(*trIt);
+
+    // overlap: It holds those keys (sequence of order n) which are common to
+    // reference corpus and translation corpus.
+    std::map<WordVector, size_t> overlap;
+    for (auto it = mergedRefNGramCounts.cbegin();
+         it != mergedRefNGramCounts.cend();
+         ++it)
+    {
+      auto translationIt = translationNGramCounts.find(it->first);
+      if (translationIt != translationNGramCounts.end())
+      {
+        // If the key (sequence of order n) is present in both translation
+        // corpus as well as reference corpus, then the minimum number of
+        // counts it has occurred in any is considered.
+        overlap[it->first] = std::min(translationIt->second, it->second);
+      }
+    }
```

Comment:
You replace 2 times `mergedRefNGramCounts` by `translationNGramCounts` and vice versa. I think both variants are correct:) However, I guess it's faster to loop over the smallest container. Does `mergedRefNGramCounts` usually smaller than `translationNGramCounts`?

Reply:
I'm not sure but I felt that `mergedRefNGramCounts` may be larger because it comes out of multiple documents/references while `translationNGramCounts` come from the predicted/generated text (you can see the parameter description for `referenceCorpus` and `translationCorpus`). I will clear this today with some benchmarking on slightly larger tests. (Though I feel what I'm thinking is correct).

Reply:
Yes! In fact, iterating over `translationNGramCounts` is faster. I got the following results (in the current test sets)--

When iterating over `mergedNGramCounts`-
```
Time taken (without smoothing function): 0.000766
Time taken (with smoothing function): 0.001878
```

While iterating over `translationNGramCounts`-
```
Time taken (without smoothing function): 0.000339
Time taken (with smoothing function): 0.000355
```
There is a significant difference even on small test. I don't think we need to do any further tests.

Reply:
Yes, I agree.